### DISCUSSION TITLE
[SSR] Use Recoil as state management

### DISCUSSION THREAD

**Discussion by TristanPouliquen** [tauri-apps/tauri#6188] [Q&A] [answered]
Hello,

I am trying to build my first Tauri App and I am trying to use Recoil as my state management.

I defined my first atoms and used them in some components.

But when I launch my Tauri App I get an error stating:

Server Error
ReferenceError: window is not defined
This error happened while generating the page. Any console logs will be displayed in the terminal window.

From what I understand, this would come from the SSR of NextJS, but this started appearing when I added Recoil to my project.

Are there specific instructions on how to implement Recoil in a NextJS project for correct SSR?

Here are the atoms & actions I defined: https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/_state/positions.ts

Here are the relevant components:

- Layout (including RecoilRoot): https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/components/layout.tsx
- CardSmall: https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/components/card_small.tsx
- Index page: https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/pages/index.tsx
- action file: https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/_state/positions.ts

This page just fetches the list of Position objects from my backend, displays them & handles a focus to open a modal when clicking on an element.

This seems to be caused by a call to `@tauri-apps/api` during the SSR but I don't understand where. All `invoke` calls come from my custom hook `usePositionActions`, I tried to look into loading dynamically but it does not seem to be possible for a function.

Thanks in advance if anyone has an idea!

**Comment by FabianLars** [comment, marked as answer]
I don't think this is necessarily Recoil's fault (well, i never heard of that lib so maybe it is) but maybe your usage of tauri's apis (`invoke` in this case). All of tauri's apis need access to the `window` and some apis need the `navigator` too.

This basically means that you need to make sure that tauri's apis are only called on the client side (for example inside an componentDidMount-like useEffect hook).

Edit: whoops, didn't see your last message where you basically said the same already, sorry...

**Reply by TristanPouliquen** [reply]
Yes, so I think there is something that makes tauri code load in the SSR mode, but I don't really understand why & where.

Debugging info is void so it doesn't help 😅 

I'm sure it is just a stupid mistake, maybe I separated my files in a way that I shouldn't have. This is my first project with Tauri, Next & Recoil so there's things I might miss 😅

**Reply by TristanPouliquen** [reply]
I updated my answer so the links I had put are clickable. I am using `invoke` in functions with the await/async keywords, but these are simple Typescript files loaded as dependencies in my components.

**Reply by FabianLars** [reply]
Sooo, i guess it should be this line then https://github.com/TristanPouliquen/selection-poste-enm/blob/state-management/pages/index.tsx#L20. i think it will run on the server-side too.

You could wrap that in a `if (window) {}` or something i guess. Or by converting it to a componentDidMount hook depending on what exactly its purpose is.

**Reply by TristanPouliquen** [reply]
Hum, I did some tests and :
- even if I remove code linked to my `usePositionActions` hook & the `useEffect` linked to it, I still get the `window is not defined` error.

I had added the code in the `useEffect` which is similar to a `componentDidMount` (see https://dev.to/vvo/how-to-solve-window-is-not-defined-errors-in-react-and-next-js-5f97) on purpose to have it run only when the component is mounted, but the problem seems to come from somewhere else.

I've found this issue with NextJs (https://github.com/facebookexperimental/Recoil/issues/2082) but the error message is different.

On my latest tests, removing all references to `invoke` method and reputting them one by one, I seem to have broken NextJS as it does not error but just does not load anything... 😅 

I'll give it a good night's rest to see, and hope someone has already faced a similar issue.

Thank you for your time!

**Reply by TristanPouliquen** [reply]
I tried some things again.

I first removed all usages of the `invoke` function from my code. This allowed me to restart my app normally as intended.

I then restored my file `_state/actions.ts` containing my hook in which I have my methods calling `invoke`.

Without even using this hook in my components, I get my error again. I don't understand at what point NextJs is indeed executing `@tauri-app/api` code to get to the error.

The error comes from this minified statement from a chunk of `@tauri-apps/api`.
```
"__TAURI_METADATA__"in window?b=new s(window.__TAURI_METADATA__.__currentWindow.label,{skip:!0}):(console.warn(`Could not find "window.__TAURI_METADATA__". The "appWindow" value will reference the "main" window label.
Note that this is not an issue if running this frontend on a browser instead of a Tauri window.`),b=new s("main",{skip:!0}));
```

I can't seem to find the file in Github that corresponds to this error.

I will continue to look into why NextJS seems to be executing code that isn't even used.

**Reply by TristanPouliquen** [reply]
I might just end up stopping using NextJS for this project, as I don't really need SSR and it seems to add complexity I do not know how to handle 😅

**Reply by TristanPouliquen** [reply]
The problem comes from https://github.com/tauri-apps/tauri/blob/dev/tooling/api/src/window.ts#L2033 thtat is apparently run at the top level of the module, which get run on import.

Would it be possible to change it to `if (typeof window !== "undefined" && '__TAURI_METADATA__' in window)` ?

**Reply by FabianLars** [reply]
That code is only executed when you import the `window` module though (which you don't as far as i can see). If that'd be the problem (without importing the window module) then we would have many more complains about nextjs in tauri.

What i'm thinking is that for whatever reason, using Recoil makes nextjs try to mount the component at build time and therefore triggering the `invoke` call, which indeed does need access to `window` too (it will fail at build time even if it has access to the window because the api on the window is only injected in an actual tauri app).

**Reply by TristanPouliquen** [reply]
Hi, I've continued trying to isolate the project. I also talked with one of the maintainers at NextJs.

What we found is that the error appears when I import my `@/_state/actions` file, which in turn imports `import {invoke} from '@tauri-apps/api`.

This file then imports `import * as window from './window'` which causes the error I linked earlier.

See his post here: https://github.com/vercel/next.js/discussions/45605#discussioncomment-4886711

This fails even if I am not using Recoil, as you can see in the two latest commits of this branch: https://github.com/TristanPouliquen/selection-poste-enm/commits/state-management

Just adding an import statement to my actions file makes my application break on window undefined.

**Reply by FabianLars** [reply]
can you try importing invoke like this then? `import {invoke} from '@tauri-apps/api/tauri` (i could swear that's what you did, i can't trust my eyes anymore 😂 )

**Reply by TristanPouliquen** [reply]
Oh, this actually worked indeed!

And it is what is in the Getting started docs actually, sorry for my mistake! 

🎉
